core/data_stream.py

```diff
@@ -0,0 +1,249 @@
+import abc
+import tempfile
+import os
+import typing
+from core.data_block import DataBlock
+
+Symbol = typing.Any
+
+
+class DataStream(abc.ABC):
+    """abstract class to represent a Data Stream
+
+    The DataStream facilitates the block interface.
+    From the interface standpoint, the two functions which are useful are:
+    - get_block(block_size) -> returns a DataBlock of the given block_size from the stream
+    - write_block(block) -> writes the block of data to the stream
+
+    The DataStream can act as a stream object for both writing and reading blocks
+    The two more useful sub-classes of the abstract class are FileDataStream and ListDataStream.
+    (see their description for more details)
+    """
+
+    @abc.abstractmethod
+    def reset(self):
```

Comment:
TODO: explain the meaning of reset in more detail. In particular, if we reset a writeable stream what happens. For lists, it seems we move back the pointer but still keep appending to the end. Whereas for file, after reset we will overwrite.

Reply:
- The difference in implementation of reset for List vs File is a bug. 
- Now that I think of it "reset" doesn't make sense while writing. It probably makes sense while reading. For example: what does it mean to "reset" a file in "w" mode vs "a" mode. 

There are two ways of dealing with this: 
a. Instead of defining a `reset` function.. we define a `seek` function to go to a particular symbol location. So `seek(0)` takes pointer to the beginning of the list/File, while `seek(k)` takes it to the location of the `kth` symbol. 
(this implementation could vary based on whether we are reading integers/characters etc.). 
This will remove the ambiguity. It also seems consistent with the default python file object which doesn't have a 'reset' method, but ahs a `seek` method. 

b) The second method is defining a separate File Writer and Reader and defining `reset` only for the Reader. 

Let me know your thoughts on either of these. I think the first option is a decent one IMO

Reply:
First option sounds good. Will seek make sense when writing? If not, we should at least add a comment to clarify.

Reply:
Done!

Reply:
Seek probably makes sense for writing, if we want to modify a particular symbol

Reply:
The code looks correct now for both list and file, thanks for adding the tests.